### PR TITLE
8295407: C2 crash: Error: ShouldNotReachHere() in multiple vector tests with -XX:-MonomorphicArrayCheck -XX:-UncommonNullCast

### PR DESCRIPTION
**Pull request by fg1417**
For unsupported `CMove` patterns, [JDK-8293833](https://bugs.openjdk.org/browse/JDK-8295407) helps remove unused `CMove` and related packs from superword candidate packset by the function `remove_cmove_and_related_packs()`, but it only works when `-XX:+UseVectorCmov` is enabled[1]. When the option is not enabled, these unsupported `CMove` packs are still kept in the superword packset, causing the same failure.

Actually, the function `filter_packs()` in superword is to filter out unsupported packs but it can't work as expected currently for these `CMove` cases. As we know, not all `CMove` packs can be vectorized. `merge_packs_to_cmovd()`[2] looks through all packs in the superword packset and generates a `CMove` candidate packset to collect all qualified `CMove` packs. Hence, only `CMove` packs in the `CMove` candidate packset are our target patterns and can be vectorized. But `filter_packs()` thinks, if the `CMove` pack is in a superword packset and its vector node is implemented in the current platform, then it can be vectorized. Therefore, the function doesn't remove these unsupported packs.

We can adjust the function `implemented()` in the stage of `filter_packs()` to check if the current `CMove` pack is in the `CMove` candidate packset. If not, `filter_packs()` considers it not to be vectorized and then remove it. After the fix, whether
`-XX:+UseVectorCmov` is enabled or not, these unsupported packs can be removed by `filter_packs()`. In this way, we don't need the function`remove_cmove_and_related_packs()` anymore and thus the patch also cleans related code.

[1] https://github.com/openjdk/jdk/blob/9b9be88bcaa35c89b6915ff0c251e5a04b10b330/src/hotspot/share/opto/superword.cpp#L537
[2] https://github.com/openjdk/jdk/blob/9b9be88bcaa35c89b6915ff0c251e5a04b10b330/src/hotspot/share/opto/superword.cpp#L1892

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295407](https://bugs.openjdk.org/browse/JDK-8295407): C2 crash: Error: ShouldNotReachHere() in multiple vector tests with -XX:-MonomorphicArrayCheck -XX:-UncommonNullCast

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11034/head:pull/11034` \
`$ git checkout pull/11034`

Update a local copy of the PR: \
`$ git checkout pull/11034` \
`$ git pull https://git.openjdk.org/jdk pull/11034/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11034`

View PR using the GUI difftool: \
`$ git pr show -t 11034`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11034.diff">https://git.openjdk.org/jdk/pull/11034.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/compiler/loopopts/TestUnsupportedConditionalMove.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8295407
+ * @summary C2 crash: Error: ShouldNotReachHere() in multiple vector tests with
+ *          -XX:-MonomorphicArrayCheck -XX:-UncommonNullCast
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/loopopts/TestUnsupportedConditionalMove.java:28]: `MonomorphicArrayCheck` and `UncommonNullCast` are debug flags, the test will fail with a release build.

**Review by TobiHartmann — COMMENTED** [review]
Looks good to me otherwise but someone more familiar with superword should look at this as well.

File: test/hotspot/jtreg/compiler/loopopts/TestUnsupportedConditionalMove.java (hunk 1 above)
Comment (fg1417) [reply test/hotspot/jtreg/compiler/loopopts/TestUnsupportedConditionalMove.java:28]: @TobiHartmann, thanks for your review! The options added here is commented as part of summary title, not as JVM options. I suppose it should be fine for a release build, right? :-)

Comment (TobiHartmann) [reply test/hotspot/jtreg/compiler/loopopts/TestUnsupportedConditionalMove.java:28]: Right, I missed that. Does the test reproduce the issue without these flags? In any case, I think a more descriptive summary would be good.

**Review by vnkozlov — CHANGES_REQUESTED** [review]
I have request since you are touching this code.
[8192846](https://bugs.openjdk.org/browse/JDK-8192846) changes were a little sloppy and did not rename some methods which causing confusion.
Please, rename `is_CmpD_candidate`, `merge_packs_to_cmpd` and others you find to general `fp` as you did with `is_cmove_fo_opcode`.

**Comment by vnkozlov** [comment]
Or may be simply remove `D`: `is_Cmp_candidate(), merge_packs_to_cmove(), test_cmp_pack()`.
There are also comments which describes only `CMoveD`.

**Comment by vnkozlov** [comment]
Do we have IR tests  to verify cmove vectorization?

File: test/hotspot/jtreg/compiler/loopopts/TestUnsupportedConditionalMove.java (hunk 1 above)
Comment (fg1417) [reply test/hotspot/jtreg/compiler/loopopts/TestUnsupportedConditionalMove.java]: Yes, the added testcase here can reproduce the issue even without these flags and I suppose those flags help C2 vectorize more hot loops in the reported testcases. I updated the summary part in the new commit. Thanks for your suggestion @TobiHartmann!

**Review by vnkozlov — COMMENTED** [review]
Looks good. I submitted testing.

File: test/hotspot/jtreg/compiler/loopopts/TestUnsupportedConditionalMove.java (hunk 1 above)
Comment (TobiHartmann) [reply test/hotspot/jtreg/compiler/loopopts/TestUnsupportedConditionalMove.java]: Great, thanks for confirming and updating the summary!

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. Let's wait for Vladimir's testing to finish.

**Comment by vnkozlov** [comment]
I got 2 failures. Description in the bug report. One could be not related. But Vector512ConversionTests.java test failure with flags from the bug report is suspicious.

**Comment by vnkozlov** [comment]
Digging in JBS shows that Vector512ConversionTests.java failure looks like [JDK-8276064](https://bugs.openjdk.org/browse/JDK-8276064). So it is not new failure.

Looks like your changes are fine.

**Review by vnkozlov — APPROVED** [review]

**Comment by fg1417** [comment]
Thanks for your kind review and test work, @TobiHartmann @vnkozlov. I'll integrate it.

/integrate

**Pull request closed without merging**